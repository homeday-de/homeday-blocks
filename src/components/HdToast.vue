<template>
<div class="toast" :class="computedClasses">
  <p class="toast__text" >
    <slot></slot>
  </p>
  <section v-if="hasLabels" class="toast__controls">
    <button
      v-if="secondaryLabel"
      class="toast__control toast__control--secondary"
      @click="secondaryClick"
      v-text="secondaryLabel"
    />
    <button
      v-if="primaryLabel"
      class="toast__control toast__control--primary"
      @click="primaryClick"
      v-text="primaryLabel"
    />
  </section>
</div>
</template>

<script>
export default {
  name: 'HdToast',
  props: {
    secondaryLabel: String,
    primaryLabel: String,
  },
  data: () => ({
    isOpen: false,
    isClosing: false,
  }),
  mounted() {
    this.$on('open', this.open);
  },
  computed: {
    hasLabels() {
      return Boolean(this.primaryLabel || this.secondaryLabel);
    },
    computedClasses() {
      return {
        'toast--isOpen': this.isOpen,
        'toast--isClosing': this.isClosing,
      };
    },
  },
  methods: {
    open() {
      this.isOpen = true;
    },
    close() {
      this.isClosing = true;
      setTimeout(() => {
        this.isOpen = false;
        this.isClosing = false;
        this.$emit('close');
      }, 500);
    },
    primaryClick() {
      this.$emit('primaryClick');
    },
    secondaryClick() {
      this.$emit('secondaryClick');
      this.close();
    },
  },
};
</script>

<style lang="scss" scoped>
@import 'hd-blocks/styles/mixins.scss';
$toastWidth: 288px;

.toast {
  // Position
  display: none;
  position: fixed;
  top: 50%;
  bottom: auto;
  transform: translateY(-50%);
  z-index: 99;
  left: 0;
  right: 0;
  margin: auto;

  @media (min-width: $break-tablet) {
    transform: none;
    left: auto;
    right: $inline-m;
    top: $stack-m;
  }

  @media (min-width: $break-tablet) {
    right: $inline-l;
    top: $stack-m;
  }

  // Appereance
  width: $toastWidth;
  background: $white;
  border-radius: $default-border-radius;
  border: 1px solid #E1E1E1;
  box-shadow: 0 2px 2px 0 rgba(0,0,0,0.25);
  @include font('xsheadline');

  &--isOpen {
    display: block;
    animation: slide-in-top 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
  }
  &--isClosing {
    animation: slide-out-top 0.45s cubic-bezier(0.755, 0.050, 0.855, 0.060) both;
  }
  &__text {
    padding: $stack-s $inline-m;

    border-bottom: 1px solid #E1E1E1;
  }
  &__controls {
    display: flex;
    flex-grow: 1;
  }

  &__control {
    @include font('xsheadline');
    width: 100%;
    height: 42px;
    border: none;
    outline: none;
    text-align: center;
    background: $white;

    &--primary {
      background: $vivid-blue;
      color: $white;
      position: relative;
      cursor: pointer;
      transition: all $time-s ease-in-out;

      // Fake the blue border
      &:after {
        content: "";
        background: transparent;
        width: 100%;
        position: absolute;
        height: 100%;
        top: 0;
        left: 0;
        margin-left: -1px;
        border-right: 1px solid $vivid-blue;
        border-left: 1px solid $vivid-blue;
        border-bottom: 1px solid $vivid-blue;

        border-bottom-right-radius: $default-border-radius;
        border-bottom-right-radius: $default-border-radius;
      }

      &:hover {
        background-color: darken($vivid-blue, 15);

        &:after {
          border-color: darken($vivid-blue, 15);
        }
      }
    }

    &--secondary {
      color: $regent-gray;
    }
  }
}


/* ----------------------------------------------
 * Generated by Animista on 2018-8-16 17:47:24
 * w: http://animista.net, t: @cssanimista
 * ---------------------------------------------- */

/**
 * ----------------------------------------
 * animation slide-in-elliptic-top-fwd
 * ----------------------------------------
 */
 @keyframes slide-in-top {
  0% {
    transform: translateY(-1000px);
    opacity: 0;
  }
  100% {
    transform: translateY(0);
    opacity: 1;
  }
}

@keyframes slide-out-top {
  0% {
    transform: translateY(0);
    opacity: 1;
  }
  100% {
    transform: translateY(-1000px);
    opacity: 0;
  }
}

@keyframes pulsate-fwd {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.05);
  }
  100% {
    transform: scale(1);
  }
}
</style>
