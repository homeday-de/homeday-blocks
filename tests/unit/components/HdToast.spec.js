import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdToast from '@/components/HdToast.vue';

const wrapperBuilder = wrapperFactoryBuilder(HdToast);

describe('HdToast', () => {
  it('renders component', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("'primaryLabel' text binds to primary <button>", async () => {
    const text = 'random-text';
    const wrapper = wrapperBuilder({
      props: {
        primaryLabel: text,
      },
    });

    const button = wrapper.get('.toast__control--primary');

    expect(wrapper.html()).toMatchSnapshot();
    expect(button.text()).toBe(text);
  });

  it("primary <button> calls 'primaryClick' when clicked", () => {
    const text = 'random-text';
    const primaryClickMock = jest.fn();
    const wrapper = wrapperBuilder({
      props: {
        primaryLabel: text,
      },
      methods: {
        primaryClick: primaryClickMock,
      },
    });

    const button = wrapper.get('.toast__control--primary');

    button.trigger('click');

    expect(primaryClickMock).toHaveBeenCalledTimes(1);
  });

  it("'secondaryLabel' text binds to secondary <button>", () => {
    const text = 'random-text';
    const wrapper = wrapperBuilder({
      props: {
        secondaryLabel: text,
      },
    });

    const button = wrapper.get('.toast__control--secondary');

    expect(wrapper.html()).toMatchSnapshot();
    expect(button.text()).toBe(text);
  });

  it("secondary <button> calls 'secondaryClick' when clicked", () => {
    const text = 'random-text';
    const secondaryClickMock = jest.fn();
    const wrapper = wrapperBuilder({
      props: {
        secondaryLabel: text,
      },
      methods: {
        secondaryClick: secondaryClickMock,
      },
    });

    const button = wrapper.find('.toast__control--secondary');

    button.trigger('click');

    expect(secondaryClickMock).toHaveBeenCalledTimes(1);
  });

  it('renders default slot', () => {
    const wrapper = wrapperBuilder({
      slots: {
        default: '<div id="slot"></div>',
      },
    });
    const slot = wrapper.get('#slot');

    expect(wrapper.html()).toMatchSnapshot();
    expect(slot.exists()).toBe(true);
  });

  describe('methods', () => {
    describe('open', () => {
      it('sets isOpen to true', () => {
        const wrapper = wrapperBuilder();
        const { vm } = wrapper;


        expect(vm.isOpen).toBe(false);

        vm.open();

        expect(vm.isOpen).toBe(true);
      });
    });

    describe('close', () => {
      it("sets 'isClosing' to true before timers then sets 'isOpen' and 'isClosing' to false and emits 'close'", () => {
        jest.useFakeTimers();

        const wrapper = wrapperBuilder();
        const { vm } = wrapper;

        vm.close();

        expect(vm.isClosing).toBe(true);

        jest.runAllTimers();

        expect(setTimeout).toHaveBeenCalledTimes(1);
        expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 500);

        expect(vm.isOpen).toBe(false);
        expect(vm.isClosing).toBe(false);

        expect(wrapper.emitted().close).toBeTruthy();
        expect(wrapper.emitted().close.length).toBe(1);
      });
    });
  });
});
