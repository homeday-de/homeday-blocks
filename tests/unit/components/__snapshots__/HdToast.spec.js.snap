// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdToast 'primaryLabel' text binds to primary <button> 1`] = `
<div class="toast">
  <p class="toast__text"></p>
  <section class="toast__controls">
    <!----> <button class="toast__control toast__control--primary">random-text</button></section>
</div>
`;

exports[`HdToast 'secondaryLabel' text binds to secondary <button> 1`] = `
<div class="toast">
  <p class="toast__text"></p>
  <section class="toast__controls"><button class="toast__control toast__control--secondary">random-text</button>
    <!---->
  </section>
</div>
`;

exports[`HdToast renders component 1`] = `
<div class="toast">
  <p class="toast__text"></p>
  <!---->
</div>
`;

exports[`HdToast renders default slot 1`] = `
<div class="toast">
  <p class="toast__text">
    <div id="slot"></div>
  </p>
  <!---->
</div>
`;
